api/rest/host.go

```diff
@@ -63,6 +63,214 @@ func fieldByIndex(value reflect.Value, index []int) reflect.Value {
+		} else {
+			websocketServeHTTP(ctx, r, w, result, reflect.Value{})
+		}
+		return
+	}
+	
+	isSeq, isSeq2 := isIteratorType(result.Type())
+	if isSeq || isSeq2 {
+		if strings.Contains(accept, "text/event-stream") {
+			handleIteratorSSE(ctx, r, w, result, isSeq2)
+		} else if strings.Contains(accept, "application/json") {
+			handleIteratorJSON(ctx, r, w, result, isSeq2)
+		} else {
+			handleIteratorWebSocket(ctx, r, w, result, isSeq2)
+		}
+		return
+	}
+}
+
+func handleChannelJSON(ctx context.Context, r *http.Request, w http.ResponseWriter, result reflect.Value, fn api.Function, auth api.Auth[*http.Request]) {
+	var finalResult any
+	cases := []reflect.SelectCase{
+		{Dir: reflect.SelectRecv, Chan: result},
+		{Dir: reflect.SelectRecv, Chan: reflect.ValueOf(ctx.Done())},
+	}
+	
+	if r.Method == "GET" {
+		chosen, value, ok := reflect.Select(cases)
+		if chosen == 1 || !ok {
+			w.WriteHeader(http.StatusNoContent)
+			return
+		}
+		finalResult = value.Interface()
+	} else if r.Method == "POST" {
+		var lastValue reflect.Value
+		for {
+			chosen, value, ok := reflect.Select(cases)
+			if chosen == 1 || !ok {
+				if lastValue.IsValid() {
+					finalResult = lastValue.Interface()
+				} else {
+					w.WriteHeader(http.StatusNoContent)
+					return
+				}
+				break
+			}
+			lastValue = value
+		}
+	} else {
+		http.Error(w, "method not allowed for channel endpoints", http.StatusMethodNotAllowed)
+		return
+	}
+	
+	w.Header().Set("Content-Type", "application/json")
+	encoder := contentTypes["application/json"]
+	if err := encoder.Encode(w, finalResult); err != nil {
+		handle(ctx, fn, auth, w, err)
+	}
+}
+
+func handleIteratorJSON(ctx context.Context, r *http.Request, w http.ResponseWriter, iterator reflect.Value, isSeq2 bool) {
+	var results []any
+	
+	if isSeq2 {
+		yieldFunc := reflect.MakeFunc(
+			iterator.Type().In(0),
+			func(args []reflect.Value) []reflect.Value {
+				select {
+				case <-ctx.Done():
+					return []reflect.Value{reflect.ValueOf(false)}
+				default:
+					pair := []any{args[0].Interface(), args[1].Interface()}
+					results = append(results, pair)
+					return []reflect.Value{reflect.ValueOf(true)}
+				}
+			},
+		)
+		iterator.Call([]reflect.Value{yieldFunc})
+	} else {
+		yieldFunc := reflect.MakeFunc(
+			iterator.Type().In(0),
+			func(args []reflect.Value) []reflect.Value {
+				select {
+				case <-ctx.Done():
+					return []reflect.Value{reflect.ValueOf(false)}
+				default:
+					results = append(results, args[0].Interface())
+					return []reflect.Value{reflect.ValueOf(true)}
+				}
+			},
+		)
+		iterator.Call([]reflect.Value{yieldFunc})
+	}
+	
+	w.Header().Set("Content-Type", "application/json")
+	encoder := contentTypes["application/json"]
+	if err := encoder.Encode(w, results); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+}
+
+func handleIteratorSSE(ctx context.Context, r *http.Request, w http.ResponseWriter, iterator reflect.Value, isSeq2 bool) {
```

Comment:
I notice we have lots of functions for different combinations of streaming, can we figure out a way to compose different aspects together using functions, parameterisation, instead of  a new function for each combination.